File: tools/regenerate-data.js
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  extractEmbeddedData,
  serializeEmbeddedData,
  validateEmbeddedAnimeShape
} from './lib/embedded-data.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const DEFAULT_INPUT = path.join(__dirname, '..', 'data', 'anime.full.json');
const DEFAULT_OUTPUT = path.join(__dirname, '..', 'js', 'data.js');

const parseArgs = (args) => {
  const values = {};
  for (let index = 0; index < args.length; index += 1) {
    const arg = args[index];
    if (!arg.startsWith('--')) continue;
    const key = arg.replace(/^--/, '');
    if (index + 1 < args.length && !args[index + 1].startsWith('--')) {
      values[key] = args[index + 1];
      index += 1;
    } else {
      values[key] = 'true';
    }
  }
  return values;
};

const ensureDir = (filePath) => {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
};

const main = () => {
  const args = parseArgs(process.argv.slice(2));
  const inputPath = args.input ? path.resolve(process.cwd(), args.input) : DEFAULT_INPUT;
  const outputPath = args.output ? path.resolve(process.cwd(), args.output) : DEFAULT_OUTPUT;

  if (!fs.existsSync(inputPath)) {
    throw new Error(`Input JSON not found: ${inputPath}`);
  }

  const payload = JSON.parse(fs.readFileSync(inputPath, 'utf8'));
  const shape = validateEmbeddedAnimeShape(payload, { sampleSize: 50 });
  if (!shape.valid) {
    throw new Error(`Input payload shape invalid:\n- ${shape.errors.join('\n- ')}`);
  }

  const script = serializeEmbeddedData(payload);
  ensureDir(outputPath);
  fs.writeFileSync(outputPath, script, 'utf8');

  // Re-parse written output to catch serialization regressions.
  const parsed = extractEmbeddedData(fs.readFileSync(outputPath, 'utf8'));
  const parsedShape = validateEmbeddedAnimeShape(parsed, { sampleSize: 50 });
  if (!parsedShape.valid) {
    throw new Error(`Generated output is invalid:\n- ${parsedShape.errors.join('\n- ')}`);
  }

  console.log(`Wrote embedded data to ${path.relative(process.cwd(), outputPath)}`);
};

try {
  main();
} catch (error) {
  console.error(error.message || error);
  process.exitCode = 1;
}

File: tools/regenerate_data.py
#!/usr/bin/env python3
"""Regenerate js/data.js-compatible embedded fallback data."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
import sys
from typing import Optional, List

from embedded_data import extract_embedded_data, serialize_embedded_data, validate_embedded_anime_shape


ROOT = Path(__file__).resolve().parents[1]
DEFAULT_INPUT = ROOT / "data" / "anime.preview.json"
DEFAULT_OUTPUT = ROOT / "js" / "data.js"


def regenerate(input_path: Path = DEFAULT_INPUT, output_path: Path = DEFAULT_OUTPUT) -> None:
    if not input_path.exists():
        raise FileNotFoundError(f"Input JSON not found: {input_path}")
    payload = json.loads(input_path.read_text(encoding="utf-8"))
    shape = validate_embedded_anime_shape(payload, sample_size=50)
    if not shape["valid"]:
        raise ValueError("Input payload shape invalid:\n- " + "\n- ".join(shape["errors"]))

    script = serialize_embedded_data(payload)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(script, encoding="utf-8")

    parsed = extract_embedded_data(output_path.read_text(encoding="utf-8"))
    parsed_shape = validate_embedded_anime_shape(parsed, sample_size=50)
    if not parsed_shape["valid"]:
        raise ValueError("Generated output is invalid:\n- " + "\n- ".join(parsed_shape["errors"]))

    try:
        label = output_path.relative_to(Path.cwd())
    except ValueError:
        label = output_path
    print(f"Wrote embedded data to {label}")


def main(argv: Optional[List[str]] = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--input", default=str(DEFAULT_INPUT))
    parser.add_argument("--output", default=str(DEFAULT_OUTPUT))
    args = parser.parse_args(argv)
    try:
        regenerate(Path(args.input), Path(args.output))
    except Exception as error:
        print(str(error), file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    sys.exit(main())
